docs/DafnyRef/integration- rust/IntegrationRust.md

```diff
@@ -71,42 +71,114 @@ For maximum flexibility, Dafny follows these rules to encode types in Rust:
+| `c.x := 2` if `c: C`          | `md!(c).x = 2`          (*) |
+| `a[0]` if `a: array<T>`       | `rd!(a)[0]`             (*) |
+| `a[1] := 2` if  `a: array<T>` | `md!(a)[1] = 2`         (*) |
+
+
+## Encoding of classes and arrays with the option --raw-pointers
+The option `--raw-pointers` ensures that classes only have 8 bytes pointers,
+and array only have 8 bytes pointers + information about the length,
+and point to raw memory. Therefore, pointers have copy semantics and
+no call to any clone method needs to be made.
+However, since pointer comparison in Rust normally compares vtables,
+which is brittle, Dafny use its own pointer types which are wrappers:
+```rs
+pub struct Ptr<T: ?Sized>(pub Option<NonNull<UnsafeCell<T>>>);
+```
 
-## With the option --raw-pointers=true
-https://doc.rust-lang.org/book/appendix-07-nightly-rust.html
 |-------------------------------|-----------------------------|
 |  Dafny type                   |   Rust type                 |
 |-------------------------------|-----------------------------|
-| `C`, `C?` (for class, iterator C) | `*mut C`                |
-| (trait) `T`                   | (trait) `*mut dyn T`        |
-| `array<T>`                    | `*mut [T]`                  |
-| `array2<T>`                   | `*mut array2<T>`            |
+| `C`, `C?` (for class, iterator C) | `Ptr<C>`            (*) |
+| (trait) `T`                   | (trait) `Ptr<dyn T>`    (*) |
+| `array<T>`                    | `Ptr<[T]>`              (*) |
+| `array2<T>`                   | `Ptr<array2<T>>`    (*) (*) |
+| `c.x` if `c: C`               | `read!(c).x`            (*) |
+| `c.x := 2` if `c: C`          | `modify!(c).x = 2`      (*) |
+| `a[0]` if `a: array<T>`       | `read!(a)[0]`           (*) |
+| `a[1] := 2` if  `a: array<T>` | `modify!(a)[1] = 2`     (*) |
 
 (*) Defined in the Dafny runtime
 
+# Externs
+
+You can provide additional `*.rs` files to `dafny translate`, `dafny build` and even `dafny run` (via the `--input` option)
+All the elements imported from the Rust files are going to be available from the generated code via the module `_dafny_externs` that imports everything.
+
+The best way to see what you have to implement as an extern is to compile your code without it. You'll see that it will be necessary for the module `_dafny_extern` to use open a certain module, which you'd put in a file `external.rs` and pass it for compilation.
+
+When using simple 1-argument externs, the compilation follows Dafny conventions:
+* Static methods are static instance methods of a `pub struct _default {}` which is defined in the module itself, or must be defined externally if all static methods are labelled as externs.
+* Methods or functions with an extern attribute with 2 arguments are interpreted as follow: The first argument is a module name, and all the "." are rewritten to "::" to follow Rust's syntax. The second argument is the name of a static `pub fn` of that module.
```

Comment:
```suggestion
* Methods or functions with an extern attribute with 2 arguments are interpreted as follows: The first argument is a module name, and all the "." are rewritten to "::" to follow Rust's syntax. The second argument is the name of a static `pub fn` of that module.
```

Also, is there a need for mapping `.` to `::`, as opposed to encouraging the extern attributes to use the Rust name more directly? We are explicitly moving away from attempting to support the same extern attribute values across multiple target languages.

Reply:
Using `.` instead of `::` makes your externs compatible with all backends. If one has a Rust-only project, they might want to use `::` to remember it's a Rust function, and that should be ok.

I rephrased the entire paragraph into this:

> 
When using simple 1-argument externs, the compilation follows Dafny conventions: static methods are static instance methods of a `pub struct _default {}` which is defined in the module itself, or must be defined externally if all static methods are labelled as externs.
When using 2-argument extern methods or functions, the compilation follows Rust convention: The first argument is a module name, and all the "." are rewritten to "::" to map into Rust's syntax The second argument is the name of a static `pub fn` of that module.

Reply:
It's not possible to make externs compatible with all backends - Go doesn't support the period separator, and there isn't a natural way to map a string like A.B.C to a Go import unambiguously.

We're also in the process of changing smithy-dafny to emit different extern declarations per language.

We should just support the idiomatic `A::B::C` form for Rust externs, and avoid confusing customers by implying one form of extern should work for all backends.

Reply:
> It's not possible to make externs compatible with all backends - Go doesn't support the period separator, and there isn't a natural way to map a string like A.B.C to a Go import unambiguously.
> We're also in the process of changing smithy-dafny to emit different extern declarations per language.

Good, that makes sense because you support Go,

> We should just support the idiomatic A::B::C form for Rust externs, and avoid confusing customers by implying one form of extern should work for all backends.

Got your point. I'll make sure I'm not mentioning "all backends".
Getting back to this paragraph, this file the `IntegrationRust.md` file, it's not the `IntegrationAllBackends` file, so everything there is not implying anything else for any other backends, it's very Rust specific.
Still, I can imagine that some people would start to create a Dafny project to compile to Rust and decide how to encode backends, in that case, I agree they should prefer writing "::" instead of ".". So I'm changing the paragraph into the following:

> When using 2-argument extern methods or functions, the compilation follows Rust convention: The first **argument is a path of Rust module names separated by `::` (or `.` for cross-compilation with e.g. the Java compiler)**. The second argument is the name of a static `pub fn` of that module.

Would that be ok for you ?

Reply:
That's definitely much less worrying, thanks!